File: algorithms_c/sort/comb.c
/* Source: https://github.com/diptangsu/Sorting-Algorithms/blob/master/C/CombSort.c */
#include "comb.h"
static int newgap(int gap)
{
    gap = (gap * 10) / 13;
    if (gap == 9 || gap == 10)
        gap = 11;
    if (gap < 1)
        gap = 1;
    return gap;
}

void comb_sort(long a[], int aSize)
{
    int gap = aSize;
    long temp;
    int i;
    for (;;)
    {
        gap = newgap(gap);
        int swapped = 0;
        for (i = 0; i < aSize - gap; i++)
        {
            int j = i + gap;
            if (a[i] > a[j])
            {
                temp = a[i];
                a[i] = a[j];
                a[j] = temp;
                swapped = 1;
            }
        }
        if (gap == 1 && !swapped)
            break;
    }
}

File: algorithms_java/sort/Comb.java
/*Source: https://github.com/TheAlgorithms/Java/blob/master/Sorts/CombSort.java */
public class Comb{
	private static boolean less(Comparable v, Comparable w){
		return (v.compareTo(w) < 0);
	}

	private static boolean exch(Comparable[] a, int i, int j){
		Comparable swap = a[i];
		a[i] = a[j];
		a[j] = swap;
		return true;
	}
 	private static int nextGap(int gap) {
    	// Shrink gap by Shrink factor
    	gap = (gap * 10) / 13;
    	return (gap < 1) ? 1 : gap;
  	}

	public static <T extends Comparable<T>> T[] combSort(T[] arr) {
    	int size = arr.length;

    	// initialize gap
    	int gap = size;

    	// Initialize swapped as true to make sure that loop runs
    	boolean swapped = true;

    	// Keep running while gap is more than 1 and last iteration caused a swap
    	while (gap != 1 || swapped) {
      		// Find next gap
      		gap = nextGap(gap);

      		// Initialize swapped as false so that we can check if swap happened or not
      		swapped = false;

      		// Compare all elements with current gap
      		for (int i = 0; i < size - gap; i++) {
        		if (less(arr[i + gap], arr[i])) {
          			// Swap arr[i] and arr[i+gap]
          			swapped = exch(arr, i, i + gap);
        		}
      		}
    	} 
	return arr;
  }
}
